cmd/thanos/receive.go

```diff
@@ -321,23 +359,78 @@ func runReceive(
 
 		s := shipper.New(logger, reg, dataDir, bkt, func() labels.Labels { return lset }, metadata.ReceiveSource)
 
-		ctx, cancel := context.WithCancel(context.Background())
-		g.Add(func() error {
-			defer runutil.CloseWithLogOnErr(logger, bkt, "bucket client")
+		// Before starting, ensure any old blocks are uploaded.
+		if uploaded, err := s.Sync(context.Background()); err != nil {
```

Comment:
The whole idea is that we want this to block so nothing is ingested before old blocks are flushed and uploaded.

Reply:
don't you think that it will cause issues when you try to exit the receiver in the middle of an upload and it blocks here?

Reply:
I already had this discussion with @brancz on this PR, but I'm not sure where GitHub put that comment history :/ . The original implementation had a 10-minute timeout on the context and the suggestion was to remove that. Let's cover that question in a follow rather than go back and forth.